File: examples/prettier-render/renderer.go
package main

import (
	"fmt"
	"os"
	"strings"

	common "github.com/dmclink/flash-cli/gen/go/common/v1"
	"golang.org/x/term"
)

// UI Palette Color Codes
const (
	Reset  = "\033[0m"
	Bold   = "\033[1m"
	Cyan   = "\033[36m"
	Green  = "\033[32m"
	Dim    = "\033[2m"
	BgLine = "\033[48;5;235m"
)

// some static metadata variables that are sent on plugin connect
var startupBanner = `
   ___          _   _   _            __                _           
  / _ \_ __ ___| |_| |_(_) ___ _ __ /__\ ___ _ __   __| | ___ _ __ 
 / /_)/ '__/ _ \ __| __| |/ _ \ '__/ \/// _ \ '_ \ / _| |/ _ \ '__|
/ ___/| | |  __/ |_| |_| |  __/ | / _  \  __/ | | | (_| |  __/ |   
\/    |_|  \___|\__|\__|_|\___|_| \/ \_/\___|_| |_|\__,_|\___|_|   
	`

var (
	instructionFront = "Enter to flip, q to quit!"
	instructionBack  = "Enter to next, q to quit!"
)

// DESIGN CONTROLLER:
// Edit this struct and its 'RenderCard' method to customize your layouts.
// This method is called sequentially by main.go every time a card is drawn.
type PrettierRenderer struct{}

// RenderCard processes a card and outputs your terminal styling format designs.
// Returns exactly three strings: (front_view, back_view, progress_bar)
func (r *PrettierRenderer) RenderCard(card *common.Flashcard, cardNum, totalCards int32, modifiers []string) (string, string, string) {
	// Query user's current terminal width dynamically to keep split grids responsive
	termWidth, _, err := term.GetSize(int(os.Stdout.Fd()))
	if err != nil || termWidth < 40 {
		termWidth = 80 // Safe fallback scale boundary width
	}

	targetGridWidth := (termWidth * 3) / 4
	if targetGridWidth < 60 {
		targetGridWidth = 60 // Enforce a sane minimum horizontal scale
	}

	// 1. Generate full layout structures for both presentation states
	frontView := generateSplitView(card, false, targetGridWidth)
	// NOTE: backView likely also contains the full layout of frontView
	// unless you don't want the question portion of the card visible after user "flips"
	backView := generateSplitView(card, true, targetGridWidth)

	// 2. Format a matching text-based timeline progress block
	// feel free to skip this or return empty "" to fall back to a
	// default progress indicator like [1/5] or pass a non empty
	// string like " " to remove progress altogether
	progressBar := fmt.Sprintf("%s[Progress Metrics: %d/%d]%s", Dim, cardNum, totalCards, Reset)

	return frontView, backView, progressBar
}

// wrapText breaks a string down into lines that perfectly fit a specific width without breaking words
func wrapText(text string, maxWidth int) []string {
	var lines []string
	paragraphs := strings.Split(text, "\n")

	for _, para := range paragraphs {
		words := strings.Fields(strings.TrimSpace(para))
		if len(words) == 0 {
			lines = append(lines, "")
			continue
		}

		currentLine := words[0]
		for _, word := range words[1:] {
			if len(currentLine)+1+len(word) <= maxWidth {
				currentLine += " " + word
			} else {
				lines = append(lines, currentLine)
				currentLine = word
			}
		}
		lines = append(lines, currentLine)
	}
	return lines
}

func generateSplitView(card *common.Flashcard, showBack bool, gridWidth int) string {
	var sb strings.Builder

	colWidth := (gridWidth / 2) - 3

	frontHeader := "FRONT (QUESTION)"
	backHeader := "BACK (ANSWER)"
	sb.WriteString(fmt.Sprintf("%s%s %-*s │ %-*s %s\n", BgLine, Bold, colWidth+1, frontHeader, colWidth+1, backHeader, Reset))

	dividerLine := strings.Repeat("─", colWidth+2)
	sb.WriteString(fmt.Sprintf("%s%s%s─┼%s%s\n", Dim, Cyan, dividerLine, dividerLine, Reset))

	frontLines := wrapText(card.Front, colWidth)
	backLines := []string{}
	if showBack {
		backLines = wrapText(card.Back, colWidth)
	}

	maxLines := len(frontLines)
	if len(backLines) > maxLines {
		maxLines = len(backLines)
	}

	for i := 0; i < maxLines; i++ {
		var left, right string
		if i < len(frontLines) {
			left = frontLines[i]
		}
		if showBack && i < len(backLines) {
			right = backLines[i]
		}

		sb.WriteString(fmt.Sprintf(" %-*s %s│%s %s%-*s%s\n",
			colWidth+1, left,
			Dim+Cyan, Reset,
			Green, colWidth+1, right, Reset,
		))
	}

	bottomLine := strings.Repeat("─", colWidth+2)
	sb.WriteString(fmt.Sprintf("%s%s%s─┴%s%s\n", Dim, Cyan, bottomLine, bottomLine, Reset))

	return sb.String()
}

File: examples/python-simple-renderer/renderer.py
import shutil

# UI Palette Constants
RESET = "\033[0m"
BOLD = "\033[1m"
WHITE = "\033[37m"
DIM = "\033[2m"
UNDERLINE = "\033[4m"

BG_CYAN = "\033[46m"
BG_GREEN = "\033[42m"

# Foreground Colors
CYAN = "\033[36m"
GREEN = "\033[32m"
YELLOW = "\033[33m"
MAGENTA = "\033[35m"

# Visual Elements
BANNER_STR = f"""
{CYAN}{BOLD}===================================================={RESET}
{MAGENTA}{BOLD}          ⚡ FLASHCARD CLI - PYTHON ENGINE ⚡          {RESET}
{CYAN}{BOLD}===================================================={RESET}
 {GREEN}➔ Layout Engine Loaded Successfully{RESET}
 {GREEN}➔ Listening for Host Commands...{RESET}
{CYAN}{BOLD}----------------------------------------------------{RESET}
"""

FRONT_INSTRUCTION = f"{YELLOW}{BOLD}[?] Guess{RESET} | {UNDERLINE}Press Enter to Flip the Card {RESET}"
BACK_INSTRUCTION  = f"{GREEN}{BOLD}[!] Nailed it{RESET} | {UNDERLINE}Press key to continue... {RESET}"

class CustomFlashcardRenderer:
    """
    Third-Party Flashcard Renderer Layout.
    Developers modify this class to customize terminal presentations.
    """
    def render_card(self, card, card_num, total_cards, unparsed_modifiers):
        """
        Processes a single flashcard and outputs custom layout formats.
        Returns a tuple of exactly three elements: (front_view, back_view, progress_bar)
        """
        # Determine active terminal window columns to scale full-width banners
        columns, _ = shutil.get_terminal_size(fallback=(80, 24))

        # 1. Format the Question State (Step A)
        front_view = self._build_banner(" FRONT SIDE (QUESTION) ", BG_CYAN, columns)
        front_view += f"\n{BOLD}{card.front}{RESET}\n\n"

        # 2. Format the Answer State (Step B)
        back_view = self._build_banner(" FRONT SIDE (QUESTION) ", BG_CYAN, columns)
        back_view += f"\n{card.front}\n\n"
        back_view += self._build_banner(" BACK SIDE (ANSWER) ", BG_GREEN, columns)
        back_view += f"\n{BOLD}{card.back}{RESET}\n\n"

        # 3. Format the Timeline Indicator (feel free to skip this step to just use default [1/4] type progress indicator)
        progress_bar = f"{DIM}[Progress Metrics: {card_num}/{total_cards}]{RESET}"

        return front_view, back_view, progress_bar

    def _build_banner(self, title_text, bg_color, total_width):
        if len(title_text) > total_width:
            title_text = title_text[:total_width-3] + "..."
            
        remaining_space = total_width - len(title_text)
        left_pad = remaining_space // 2
        right_pad = remaining_space - left_pad

        return f"{bg_color}{WHITE}{BOLD}{' ' * left_pad}{title_text}{' ' * right_pad}{RESET}\n"
